app/helpers/duties_helper.rb

```diff
@@ -42,19 +42,49 @@ def process_day_place(duties)
     result
   end
 
-  def process_duty(duties)
-    colspan = 0
+  def check_condition(span_free, span_user_id, span_request_user_id, current_duty)
+    span_free != current_duty.free ||
+      (!span_free && !current_duty.free &&
+        ((!span_request_user_id.nil? && span_request_user_id != current_duty&.request_user_id) ||
+          span_user_id != current_duty&.user&.id))
+  end
+
+  def get_index_array(duties)
+    index_array = [0]
```

Comment:
Can you see if this function can be reworked into something like this?

```
index_array = duties.length.times.each_cons(2).select do |prev, current|
  span_free = prev.free
  ...

  check_condition(...)
end.map(&:first)
```

That way
- you do not need to duplicate the `span_free` `span_user_id` `span_request_user_id`
- it is clear that you are going through in chunks of 2

Reply:
I used the each_cons idea but I am not sure about the map part, I tried it but couldn't really understand it. Will discuss with you.